src/mlpack/methods/ann/loss_functions/poisson_nll_loss_impl.hpp

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file poisson_nll_loss_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Poisson NegativeLogLikelihood class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+
+
+// In case it hasn't yet been included.
+#include "poisson_nll_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PoissonNLLLoss<InputDataType, OutputDataType>::PoissonNLLLoss(
+  const bool logInput,
+  const bool full,
+  const double eps,
+  const bool reduce):
+  logInput(logInput),
+  full(full),
+  eps(eps),
+  reduce(reduce)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double PoissonNLLLoss<InputDataType, OutputDataType>::Forward(
+  const InputType&& input,
+  TargetType&& target)
+{
+  double lossThis;
+  double totalLossReduced = 0;
+
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    if (logInput)
+    {
+      lossThis = std::exp(input[i]) - target[i] * input[i];
+    }
+    else
+    {
+      lossThis = input[i] - target[i] * std::log(input[i] + eps);
+    }
+
+    if ((full) && (target[i] > 1))
```

Comment:
Picky style issue, no need of () around full. Sorry that I missed this in previous review.

Reply:
I think that's not an issue and probably this would be the last commit for this pull request. :)

Reply:
Sure, not an issue.